### PR TITLE
metric: rocksdb statistics detail

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Collect rocksdb's internal statistics.
@siddontang @disksing PTAL

### CODE REVIEW

File: src/raftstore/store/metrics.rs [hunk 1]
```diff
@@ -136,6 +137,68 @@ lazy_static! {
             &["cf"]
         ).unwrap();
 
+    pub static ref STORE_ENGINE_CACHE_EFFICIENCY_VEC: GaugeVec =
```
Comment (siddontang) [inline src/raftstore/store/metrics.rs]: do we need a rocksdb engine metric file independently?

**Comment by siddontang** [comment]
I think maybe it is better to use another thread to report rocksdb metrics now, not in raftstore thread.

File: src/raftstore/store/msg.rs [hunk 2]
```diff
@@ -36,6 +36,7 @@ pub enum Tick {
     CompactLockCf,
     ConsistencyCheck,
     ReportRegionFlow,
+    EngineStatDetail,
```
Comment (ngaut) [inline src/raftstore/store/msg.rs]: s/EngineStatDetail/EngineStatistics/

**Comment by siddontang** [comment]
Enable rocksdb statistic by default

File: src/raftstore/store/metrics.rs (hunk 1 above)
Comment (siddontang) [inline src/raftstore/store/metrics.rs]: any update?
I suggest using a `rocksdb_metrics.rs` for all rocksdb metrics. 
Btw, I also think we should move the rocksdb metric get from `store.rs` to another file.

File: src/raftstore/store/store.rs [hunk 3]
```diff
@@ -1528,9 +1646,23 @@ impl<T: Transport, C: PdClient> Store<T, C> {
 
     fn on_pd_store_heartbeat_tick(&mut self, event_loop: &mut EventLoop<Self>) {
         self.store_heartbeat_pd();
+        self.engine_statistics_detail();
         self.register_pd_store_heartbeat_tick(event_loop);
     }
 
+    fn engine_statistics_detail(&mut self) {
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: flush_engine_statistics

File: src/raftstore/store/store.rs [hunk 4]
```diff
@@ -69,6 +70,123 @@ const ROCKSDB_CUR_SIZE_ALL_MEM_TABLES_PROPERTY: &'static str = "rocksdb.cur-size
...
+                                                     TickerType::NumberKeysRead,
+                                                     TickerType::BytesWritten,
+                                                     TickerType::BytesRead,
+                                                     TickerType::IterBytesRead,
+                                                     TickerType::StallMicros];
+
+const ENGINE_HIST_TYPES: &'static [HistType] =
+    &[HistType::DbGetMicros, HistType::DbWriteMicros, HistType::DbSeekMicros];
+
+fn engine_ticker_metrics_set(t: TickerType, value: u64) {
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: flush_engine_ticker_metrics

File: src/raftstore/store/store.rs [hunk 5]
```diff
@@ -69,6 +70,123 @@ const ROCKSDB_CUR_SIZE_ALL_MEM_TABLES_PROPERTY: &'static str = "rocksdb.cur-size
...
+            STORE_ENGINE_FLOW_VEC.with_label_values(&["iter_bytes_read"]).set(value as f64);
+        }
+        TickerType::StallMicros => {
+            STORE_ENGINE_STALL_MICROS.set(value as f64);
+        }
+        _ => {}
+    }
+}
+
+fn engine_hist_metrics_set(t: HistType, value: HistogramData) {
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: flush_engine_history_metrics

**Comment by siddontang** [comment]
any update @zhangjinpeng1987 ?

**Comment by andelf** [comment]
ref: https://prometheus.io/docs/practices/naming/

`efficiency` is misleading.

**Comment by siddontang** [comment]
Do we test how much performance reduce?

**Comment by zhangjinpeng1987** [comment]
As RocksDB's bench, statistics will introduce about 5%-10% overhead for RocksDB, for TiKV the overhead should less than that. I will bench it soon.

**Comment by zhangjinpeng1987** [comment]
As I benched, enable statistics will reduce 1% performance for select, and reduce 2-3% performance for insert.

**Comment by siddontang** [comment]
Ping @zhangjinpeng1987

**Comment by siddontang** [comment]
LGTM

PTAL @disksing

**Comment by disksing** [comment]
LGTM

**Pull request merged**